src/ListMessageSplitter.ts

```diff
@@ -0,0 +1,319 @@
+        `server ${server.name()}`,
+        `server ${server.name()}`
+      )
+    }
+
+    if (this.servers.length === 0) {
+        splitter.add_paragraph(
+            "No servers observed",
+            "No servers observed"
+        )
+    }
+
+    // Reply to an event with the whole deal, splitting into multiple messages as needed.
+    splitter.reply(mjolnir.client, roomId, event, true)
+    ```
+*/
+
+import {LogService, MatrixClient, RichReply} from "matrix-bot-sdk";
+
+// Helper type for html + text tuples.
+export type MessageSnippet = { html: string, text: string };
+
+// The max size a message can be, with 24_000 picked at random to accommodate for overhead.
+//
+// Note: This amount was checked through trial and error, a conservative estimate would be
+// 65_536 / 2, though this author does not know overhead estimates of matrix E2EE.
+const OVERHEAD = 24_000
+const MAX_SIZE = 65_536 - OVERHEAD;
+
+// The extra bits that a <ul> tag wrapping would add to a message
+const UL_TAG_WRAP_SIZE = "<ul></ul>".length;
+// The extra bits that a <il> list tag wrapping would add to a message
+const HTML_LIST_ITEM_EXTRA_SIZE = "<li></li>".length;
+// The extra bits that listification of plaintext items would add to a message.
+const TEXT_LIST_ITEM_EXTRA_SIZE = " * \n".length;
+
+// An "Item" object to push into the list splitter.
+export class MessageListItem {
+    public readonly html: string;
+    public readonly text: string;
+
+    constructor(
+        html: string,
+        text: string,
+    ) {
+        if (html.length > (MAX_SIZE / 2)) {
+            throw new Error("HTML string too long for one item")
+        } else if (text.length > (MAX_SIZE / 2)) {
+            throw new Error("text string too long for one item")
+        }
+        this.html = html;
+        this.text = text;
+    }
+
+    public size(): number {
+        return this.html.length + HTML_LIST_ITEM_EXTRA_SIZE
+            + this.text.length + TEXT_LIST_ITEM_EXTRA_SIZE
+    }
+}
+
+// A "header" object to push into the list splitter.
+export class MessageListHeader extends MessageListItem {
+    public size(): number {
+        return this.html.length + this.text.length
+    }
+}
+
+// An internal helper class to hold a series of items, together with an optional header.
+//
+// Mainly provides a coherent split_at_size function that would allow for on-demand-sized splitting of
+// listings with headers.
+class MessageListing {
+    public items: MessageListItem[] = [];
+
+    constructor(public header: MessageListHeader | null) {
+    }
+
+    // Attempts to split this listing into a `sized` and `rest` listing.
+    //
+    // Returns:
+    // `sized` != null, if sized was adequate, or had to be split
+    // `rest` != null, if sized was not adequate (and had to be split),
+    //   or the first item is too big to be split at the desired size.
+    public split_at_size(desired_size: number): {sized: MessageListing | null, rest: MessageListing | null} {
```

Comment:
style: To keep documentation consistent, we should really be following JSDoc syntax https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html. A majority of the code base does this already if you need pointers (and editors like VSCode should generate the right parameters automatically when you do `/**` _newline_)